### PR TITLE
Set BUILD_SHARED_LIBS and CMAKE_POSITION_INDEPENDENT_CODE as Cache variables in toolchain

### LINKED ISSUE #11840: [bug] CMakeToolchain: `BUILD_SHARED_LIBS` & `CMAKE_POSITION_INDEPENDENT_CODE` not honored if CMP0077 OLD (or not set) and turned as options in upstream CMakeLists
### Environment Details (include every applicable attribute)
  * Operating System+version: macOS Montery
  * Compiler+version: AppleCland 13
  * Conan version: 1.51.1
  * Python version: 3.9.13

### Steps to reproduce (Include if Applicable)

- checkout this PR of libccd recipe (conan v2 migration): https://github.com/conan-io/conan-center-index/pull/12110 (this commit: https://github.com/conan-io/conan-center-index/pull/12110/commits/14aa84b08d915dd01c1b43b599b012b3293ff79a)
- remove `tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"` from its conanfile.py
- run `conan create . libccd/2.1@ -o libccd:shared=False`

A shared lib is created instead of a static lib, while it was working fine with the conan v1 flavor of this recipe. If `CMAKE_POLICY_DEFAULT_CMP0077` is injected as a CMakeToolchain `cache_variables` (by the way this name is super misleading, since CMakeToolchain variables are CMake cache variables, and CMakeToolchain cache_variables are CMakePresets cacheVariables), static lib is properly created.

Explanation:
- In `libccd` CMakeLists:
  - `cmake_minimum_required` is < 3.13: https://github.com/danfis/libccd/blob/v2.1/CMakeLists.txt#L1
  - `BUILD_SHARED_LIBS` is overridden as an option() defaulted to ON: https://github.com/danfis/libccd/blob/v2.1/CMakeLists.txt#L23
- without https://cmake.org/cmake/help/latest/policy/CMP0077.html set to NEW, option() removes non-cached variable of the same name

### Logs (Executed commands with output) (Include/Attach if Applicable)

<details><summary>Click to expand log</summary>

```
Exporting package recipe
libccd/2.1 exports: File 'conandata.yml' found. Exporting it...
libccd/2.1 exports: Copied 1 '.yml' file: conandata.yml
libccd/2.1: The stored package has not changed
libccd/2.1: Using the exported files summary hash as the recipe revision: 754e548f10cd5c5a13b36bd666b79ac2
libccd/2.1: Exported revision: 754e548f10cd5c5a13b36bd666b79ac2
Configuration (profile_host):
[settings]
arch=x86_64
build_type=Debug
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=13
os=Macos
[options]
libccd:shared=False
[build_requires]
[env]
[conf]
tools.cmake.cmaketoolchain:generator=Ninja

Configuration (profile_build):
[settings]
arch=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=13
os=Macos
[options]
[build_requires]
[env]
[conf]
tools.cmake.cmaketoolchain:generator=Ninja

libccd/2.1: Forced build from source
libccd/2.1 (test package): Installing package
Requirements
    libccd/2.1 from 'conancenter' - Cache
Packages
    libccd/2.1:92958fbd7bb641f338e28ce02103bd2e6bd0c465 - Build

Installing (downloading, building) binaries...
libccd/2.1: Copying sources to build folder
libccd/2.1: Building your package in /Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465
libccd/2.1: Generator txt created conanbuildinfo.txt
libccd/2.1: Calling generate()
libccd/2.1: Aggregating env generators
libccd/2.1: Calling build()
libccd/2.1: CMake command: cmake -G "Ninja" -DCMAKE_TOOLCHAIN_FILE="/Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Debug" "/Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/src"
CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Using Conan toolchain: /Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/generators/conan_toolchain.cmake
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- The C compiler identification is AppleClang 13.1.6.13160021
-- The CXX compiler identification is AppleClang 13.1.6.13160021
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:23 (option):
  Policy CMP0077 is not set: option() honors normal variables.  Run "cmake
  --help-policy CMP0077" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  For compatibility with older versions of CMake, option is clearing the
  normal variable 'BUILD_SHARED_LIBS'.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Performing Test COMPILER_SUPPORTS_VISIBILITY
-- Performing Test COMPILER_SUPPORTS_VISIBILITY - Success
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/Debug
libccd/2.1: CMake command: cmake --build "/Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/Debug" '--' '-j12'
[7/7] Creating library symlink src/libccd.2.dylib src/libccd.dylib
libccd/2.1: Package '92958fbd7bb641f338e28ce02103bd2e6bd0c465' built
libccd/2.1: Build folder /Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/Debug
libccd/2.1: Generated conaninfo.txt
libccd/2.1: Generated conanbuildinfo.txt
libccd/2.1: Generating the package
libccd/2.1: Package folder /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465
libccd/2.1: Calling package()
libccd/2.1: Copied 1 file: BSD-LICENSE
libccd/2.1: CMake command: cmake --install "/Users/spaceim/.conan/data/libccd/2.1/_/_/build/92958fbd7bb641f338e28ce02103bd2e6bd0c465/build/Debug" --prefix "/Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465"
-- Install configuration: "Debug"
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/ccd/ccd-config.cmake
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/ccd/ccd-config-version.cmake
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/pkgconfig/ccd.pc
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/share/doc/ccd/BSD-LICENSE
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/libccd.2.0.dylib
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/libccd.2.dylib
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/libccd.dylib
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/ccd.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/compiler.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/ccd_export.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/quat.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/vec3.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/include/ccd/config.h
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/ccd/ccd-targets.cmake
-- Installing: /Users/spaceim/.conan/data/libccd/2.1/_/_/package/92958fbd7bb641f338e28ce02103bd2e6bd0c465/lib/ccd/ccd-targets-debug.cmake
libccd/2.1 package(): Packaged 1 file: BSD-LICENSE
libccd/2.1 package(): Packaged 6 '.h' files
libccd/2.1 package(): Packaged 3 '.dylib' files: libccd.dylib, libccd.2.0.dylib, libccd.2.dylib
libccd/2.1 package(): Packaged 1 '.cmake' file: conan-official-libccd-targets.cmake
libccd/2.1: Package '92958fbd7bb641f338e28ce02103bd2e6bd0c465' created
libccd/2.1: Created package revision 945ad782be184094cd6378b52deaa4af
```

</details>

### PR DESCRIPTION
**Pull request by jcar87**
…riables in CMakeToolchain

Changelog: Bugfix: When recipes have `shared` and `fPIC` as options, define `BUILD_SHARED_LIBS` and `CMAKE_POSITION_INDEPENDENT_CODE` as CACHE variables in the generated `cmake_toolchain.cmake` instead of regular variables, so that they are not masked by further calls to `options()`.

Docs: Omit

Close https://github.com/conan-io/conan/issues/11840

## Notes
The bug in #11840 can happen when CMake is operating on a policy version older than 3.13 and the project's CMake scripts make `option()` calls for either `BUILD_SHARED_LIBS` and `CMAKE_POSITION_INDEPENDENT_CODE`. These calls can shadow the values defined by the toolchain (if the toolchain has been loaded already by `project()`). Using `option()` for these variables seems valid as per CMake's own [documentation](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html).

These variables are only defined in the generated `conan_toolchain.cmake` if the recipe has `fPIC` and `shared` as options. These may be a tad outside of the scope of CMake toolchains.

By setting them as cache variables, we ensure that `option()` does not override their values if `option()` is called after the toolchain is loaded, but we also ensure that if the users manually and explicitly overrides the value via `-D` flag, that value will be respected too.

One caveat is that for projects that set the values after `cmake_minimum_required()` but _before_ `project()` - the behaviour remains undefined, the value will depend on how the variable is set.

### CODE REVIEW

File: conans/test/functional/toolchains/cmake/test_cmake_toolchain.py [hunk 1]
```diff
@@ -1061,3 +1061,71 @@ def package(self):
...
+    # Note: from this point forward, the CMakeCache is already initialised.
+    # When explicitly overriding `CMAKE_POSITION_INDEPENDENT_CODE` via command line, ensure
+    # this takes precedence to the value defined by the toolchain
+    t.run_command("cmake -S . -B build/ -DCMAKE_POSITION_INDEPENDENT_CODE=ON")
+    assert "mylib target type: STATIC_LIBRARY" in t.out
+    assert "mylib position independent code: ON" in t.out
+
+    t.run_command("cmake -S . -B build/ -DBUILD_SHARED_LIBS=ON")
+    assert "mylib target type: SHARED_LIBRARY" in t.out
+    assert "mylib position independent code: ON" in t.out
```
Comment (franramirez688) [inline conans/test/functional/toolchains/cmake/test_cmake_toolchain.py]:
```suggestion
    assert "mylib position independent code: ON" in t.out

```

**Review by czoido — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]

**Comment by garethsb** [comment]
Note that this also resolves the similar issue if the upstream _CMakeLists.txt_ uses:

```
set(BUILD_SHARED_LIBS ON CACHE BOOL "Build shared libraries")
```

A workaround in Conan 1.53.0 is:

```
        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0126"] = "NEW"
```

Whereas CMP077 is for `option()` variables, CMP0126 is for `set(CACHE)` ones - see https://cmake.org/cmake/help/latest/policy/CMP0126.html.

**Comment by memsharded** [comment]
Thanks for the tips @garethsb !

**Comment by garethsb** [comment]
Unfortunately, it looks like the 1.53.0 workaround wasn't effective... all packages were built, all test packages passed... but actually a shared package had been built and used in some cases. #14227 settled on a different workaround:

```diff
-        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0126"] = "NEW"
+        tc.variables["BUILD_SHARED_LIBS"] = self.options.shared
```

**Pull request merged**